src/Parsing/ParserState.php

```diff
@@ -69,15 +64,21 @@ public function __construct($text, Settings $parserSettings, $lineNumber = 1)
     /**
      * Sets the charset to be used if the CSS does not contain an `@charset` declaration.
      *
-     * @param string $charset
+     * @throws SourceException
      */
-    public function setCharset($charset): void
+    public function setCharset(string $charset): void
     {
+        // The charset property must be set for `strsplit()`, but may need to be reverted if it fails.
+        $oldCharset = $this->charset;
         $this->charset = $charset;
-        $this->characters = $this->strsplit($this->text);
-        if (\is_array($this->characters)) {
-            $this->length = \count($this->characters);
+
+        $newCharacters = $this->strsplit($this->text);
+        if (!\is_array($newCharacters)) {
```

Comment:
I don't think we need to catch this case. `$this->strsplit` is expected to return `array`s only. (And if the `preg_split` in that function fails, we should handle that there, not here.)

Reply:
#962 to ensure `strsplit()` will only ever return arrays.